extras/devices/LVFS/LVFS.py

```diff
@@ -0,0 +1,332 @@
+                        raise_stderr=False,
+                    )
+                    if rc == 0:
+                        logger.debug(stdout)
+                        reboot = True
+                    else:
+                        logger.info(
+                            "[%s] fail to force install (downgrade) firmware %s"
+                            % (dev_name, prev_ver["Uri"].split("/")[-1])
+                        )
+                        logger.debug(stdout)
+                        logger.debug(stderr)
+                else:
+                    logger.info(
+                        "[%s] unsupported Flags: %s"
+                        % (dev_name, str(prev_ver["Flags"]))
+                    )
+            else:
+                logger.info(
+                    "[%s] already the previous version of latest release"
+                    % dev_name
+                )
+        return reboot
+
+    def check_results(self):
+        """
+        Get upgrade/downgrade result and validate if it succeeds
+
+        :returns: overall upgrade status
+        :rtype: boolean
+        """
+        fwupd_result = True
+        for dev in self.fw_info:
+            try:
+                dev_name = dev["Name"]
+            except KeyError:
+                dev_name = dev["DeviceId"]
+            try:
+                expected_ver = dev["targetVersion"]
+            except KeyError:
+                continue
+            rc, stdout, stderr = self.run_cmd(
+                "sudo fwupdmgr get-results %s --json" % dev["DeviceId"]
+            )
+            get_results = json.loads(stdout)
+            logger.debug("$fwupdmgr get-result = \n%s" % stdout)
+            try:
+                new_fw = get_results["Releases"][0]
+                update_state = get_results["UpdateState"]
+            except KeyError:
+                msg = "[%s] unable to determine if new firmware is landed due to missing results"
+                logger.info(msg)
+                print(msg)
+                fwupd_result = False
+                continue
+
+            if new_fw["Version"] == expected_ver and update_state == 2:
+                msg = "[%s] firmware flashed %s → %s" % (
+                    dev_name,
+                    dev["Version"],
+                    expected_ver,
+                )
+            else:
+                FwupdUpdateState = [
+                    "FWUPD_UPDATE_STATE_UNKNOWN",
+                    "FWUPD_UPDATE_STATE_PENDING",
+                    "FWUPD_UPDATE_STATE_SUCCESS",
+                    "FWUPD_UPDATE_STATE_FAILED",
+                    "FWUPD_UPDATE_STATE_NEEDS_REBOOT",
+                    "FWUPD_UPDATE_STATE_FAILED_TRANSIENT",
+                    "FWUPD_UPDATE_STATE_LAST",
+                ]
+                msg = "[%s] firmware upgrade failed. %s" % (
+                    dev_name,
+                    FwupdUpdateState[update_state],
+                )
+                fwupd_result = False
+            print(msg)
```

Comment:
Is print here kept on purpose?

Reply:
Yes it's to provide information for actionable operation (display upgradable/downgradable firmware).